src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannel.cs

```diff
@@ -0,0 +1,427 @@
+        {
+            return RequestChannel.BeginOpen(timeout, callback, state);
+        }
+
+        /// <summary>
+        /// Begins an asynchronous operation to open a communication object.
+        /// </summary>
+        /// <param name="callback">
+        /// The <see cref="AsyncCallback" /> delegate that receives notification of the completion of the asynchronous 
+        /// close operation.
+        /// </param>
+        /// <param name="state">
+        /// An object, specified by the application, that contains state information associated with the asynchronous 
+        /// close operation.
+        /// </param>
+        /// <returns>The <see cref="IAsyncResult" /> that references the asynchronous open operation.</returns>
+        IAsyncResult ICommunicationObject.BeginOpen(AsyncCallback callback, object state)
+        {
+            return RequestChannel.BeginOpen(callback, state);
+        }
+
+        /// <summary>
+        /// Causes a communication object to transition from its current state into the closed state.
+        /// </summary>
+        /// <param name="timeout">
+        /// The <see cref="TimeSpan" /> that specifies how long the open operation has to complete before timing out.
+        /// </param>
+        void ICommunicationObject.Close(TimeSpan timeout)
+        {
+            RequestChannel.Close(timeout);
+        }
+
+        /// <summary>
+        /// Causes a communication object to transition from its current state into the closed state.
+        /// </summary>
+        void ICommunicationObject.Close()
+        {
+            RequestChannel.Close();
+        }
+
+        /// <summary>
+        /// Completes an asynchronous operation to close a communication object.
+        /// </summary>
+        /// <param name="result">The <see cref="IAsyncResult" /> that is returned by a call to the BeginClose() method.</param>
+        void ICommunicationObject.EndClose(IAsyncResult result)
+        {
+            RequestChannel.EndClose(result);
+        }
+
+        /// <summary>
+        /// Completes an asynchronous operation to open a communication object.
+        /// </summary>
+        /// <param name="result">The <see cref="IAsyncResult" /> that is returned by a call to the BeginClose() method.</param>
+        void ICommunicationObject.EndOpen(IAsyncResult result)
+        {
+            RequestChannel.EndOpen(result);
+        }
+
+        /// <summary>
+        /// Causes a communication object to transition from the created state into the opened state within a specified interval of time.
+        /// </summary>
+        /// <param name="timeout">
+        /// The <see cref="TimeSpan" /> that specifies how long the open operation has to complete before timing out.
+        /// </param>
+        void ICommunicationObject.Open(TimeSpan timeout)
+        {
+            RequestChannel.Open(timeout);
+        }
```

Comment:
If the passed in IRequestChannel is already open, this will throw. Either need a state check here and only call Open if in state Created, or add validation that the passed in channel should be in the Created state.

Reply:
When would it be called if it was already open?

Reply:
Using our code it won't be. But you've got to protect against people using the public API's outside of our usage. So someone could do this:
```
var factory = new ChannelFactory<IRequestChannel>(binding, endpoint);
var channel = factory.CreateChannel();
channel.Open();
var trustChannel = new WSTrustChannel(factory, channel, WsTrustVersion.Trust13);
((ICommunicationObject)trustChannel).Open();
```
If this code is a little more convoluted and the IRequestChannel gets passed in from somewhere else, someone could presume a newly created WSTrustChannel needs to be opened as that's the normal case. So thinking about it, I think we should throw in the constructor if the passed in IRequestChannel isn't in the Created state as that's really where the problem comes from.  
Also, while it's not part of the ICommunicationObject interface (which is explicitly implemented so you must cast to call open), do you think there's any benefit in adding an OpenAsync and CloseAsync method to this class? No cast would be needed, but this class isn't like a regular channel instance anyway so I think that's fine. Just use Task.Factory.FromAsync to wrap the Begin/End methods. Otherwise we're expecting people to either handle APM (I refer you to our teams chat about that) or use the synchronous versions.